File: indexer.py
#!/usr/bin/env python3
# -*- coding: utf8 -*-
# Copyright (c) 2020 Roberto Treviño Cervantes

#########################################################################
#                                                                       #
# This file is part of FUTURE (Powered by Monad).                       #
#                                                                       #
# FUTURE is free software: you can redistribute it and/or modify        #
# it under the terms of the GNU General Public License as published by  #
# the Free Software Foundation, either version 3 of the License, or     #
# (at your option) any later version.                                   #
#                                                                       #
# FUTURE is distributed in the hope that it will be useful,             #
# but WITHOUT ANY WARRANTY; without even the implied warranty of        #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         #
# GNU General Public License for more details.                          #
#                                                                       #
# You should have received a copy of the GNU General Public License     #
# along with FUTURE.  If not, see <https://www.gnu.org/licenses/>.      #
#                                                                       #
#########################################################################

from typing import Callable, Iterator
import scrapy, re, gensim, h5py, string, lmdb, tldextract, json
from urllib.parse import urljoin, urlparse
from scrapy.crawler import CrawlerProcess
from nltk.tokenize import word_tokenize
from gensim.models import KeyedVectors
from polyglot.detect import Detector
from Monad import *
import numpy as np

import bson

bson.loads = bson.BSON.decode
bson.dumps = bson.BSON.encode


class Vocabulary(object):
    """Simple vocabulary wrapper."""
    def __init__(self) -> None:
        self.word2idx: dict = {}
        self.idx2word: dict = {}
        self.idx: int = 0

    def addWord(self, word: str) -> None:
        if not word in self.word2idx:
            self.word2idx[word] = self.idx
            self.idx2word[self.idx] = word
            self.idx += 1

    def __call__(self, word: str) -> int:
        if not word in self.word2idx:
            return self.word2idx["<unk>"]
        return self.word2idx[word]

    def __len__(self) -> int:
        return len(self.word2idx)


from image_tagger import *


def getPropertyFromHTMLResponse(response, property: str) -> str:
    if property == "header":
        webPageProperty = response.css("h1 ::text").getall()
    elif property == "title":
        webPageProperty = response.css("title ::text").getall()
    elif property == "body":
        return " ".join(
            re.split(
                "\s+",
                u"".join(
                    response.xpath(
                        "//body/descendant-or-self::*[not(self::script)]/text()"
                    ).extract()).strip(),
                flags=re.UNICODE,
            ))
    return " ".join(
        re.split("\s+",
                 max(webPageProperty, key=len, default=""),
                 flags=re.UNICODE))


def getWebpageMeanVector(response) -> list:
    metaDescription: str = response.xpath(
        "//meta[@property='og:description']/@content").extract_first()
    if metaDescription:
        metaTitle: str = response.xpath(
            "//meta[@property='og:title']/@content").extract_first()
        if metaTitle:
            webPageTopic: str = metaTitle
        else:
            webPageHeader: str = getPropertyFromHTMLResponse(
                response, "header").strip()
            webPageTitle: str = getPropertyFromHTMLResponse(response,
                                                            "title").strip()
            webPageTopic: str = webPageHeader + ". " + webPageTitle

        return [
            getTextVectors(webPageTopic),
            metaDescription,
            Detector(metaDescription).language.name,
        ]
    else:
        webPageBody: str = getPropertyFromHTMLResponse(response,
                                                       "body").strip()
        webPageHeader: str = getPropertyFromHTMLResponse(response,
                                                         "header").strip()
        webPageTitle: str = getPropertyFromHTMLResponse(response,
                                                        "title").strip()
        wholeWebPageText: str = webPageBody + ". " + webPageHeader + ". " + webPageTitle
        return [
            getTextVectors(wholeWebPageText),
            webPageBody,
            Detector(wholeWebPageText).language.name,
        ]


class Indexer(scrapy.Spider):
    name = "indexer"
    allowed_urls = ["*"]
    custom_settings = {
        "CONCURRENT_REQUESTS": 200,
        "CONCURRENT_REQUESTS_PER_DOMAIN": 25,
        "ROBOTSTXT_OBEY": True,
        "CONCURRENT_ITEMS": 100,
        "REACTOR_THREADPOOL_MAXSIZE": 400,
        # Hides printing item dicts
        "LOG_LEVEL": "INFO",
        "RETRY_ENABLED": False,
        "REDIRECT_MAX_TIMES": 1,
        # Stops loading page after 5mb
        "DOWNLOAD_MAXSIZE": 100000000,
        # Grabs xpath before site finish loading
        "DOWNLOAD_FAIL_ON_DATALOSS": False,
        "DOWNLOAD_DELAY": 2.0,
        "AUTOTHROTTLE_ENABLED": True,
        "JOBDIR": "./indexer_state",
        # "SCHEDULER_PRIORITY_QUEUE": "scrapy.pqueues.DownloaderAwarePriorityQueue",
        "COOKIES_ENABLED": False,
        "DOWNLOAD_TIMEOUT": 30,
        "AJAXCRAWL_ENABLED": True
    }

    start_urls = ["https://techcrunch.com/"]

    def parse(self, response) -> Iterator:
        webPageVector = getWebpageMeanVector(response)
        if webPageVector[1] != "":
            webPageSummary = webPageVector[0]
            url = response.request.url
            ImageDBTransaction = images.begin(write=True)
            for id, imageLink in enumerate([
                    str(urljoin(url,
                                urlparse(url).path) + imageHTMLTagSource) if
                    imageHTMLTagSource.startswith("/") else imageHTMLTagSource
                    for imageHTMLTagSource in response.xpath(
                        "//img/@src").extract()
            ]):
                try:
                    ImageDBTransaction.put(
                        encodeURLAsNumber(imageLink, ":image:" + str(id)),
                        bson.dumps({
                            "image_vec":
                            getSentenceMeanVector(
                                tagImage(imageLink,
                                         vocabularyPickle)).tostring(),
                            "word_vec":
                            np.array([
                                wordVector for wordVector in webPageSummary[0]
                                if type(wordVector) is np.ndarray
                            ]).mean(axis=0).tostring(),
                            "url":
                            imageLink,
                        }))
                except Exception as e:
                    print(e)
            ImageDBTransaction.commit()
            for id, vector in enumerate(webPageSummary[0]):
                webPageDomain: str = response.xpath(
                    "//meta[@property='og:site_name']/@content").extract_first(
                    )
                if webPageDomain:
                    webPageDomain = webPageDomain
                else:
                    webPageDomain: str = tldextract.extract(url).domain.upper()

                FUTURE.addElementToIndex(
                    encodeURLAsNumber(url, id),
                    bson.dumps({
                        "vec": np.array(vector).tostring(),
                        "vec_id": id,
                        "language": webPageVector[2],
                        "sentence": webPageSummary[1][id],
                        "url": url,
                        "domain": webPageDomain,
                        "body": webPageVector[1],
                    }))
        for href in response.css("a::attr(href)"):
            yield response.follow(href, self.parse)


if __name__ == "__main__":
    FUTURE = Monad("future_urls")
    images = lmdb.open("future_images", map_size=int(1e12), writemap=True)

    process: Callable = CrawlerProcess({
        "USER_AGENT":
        "FUTURE by Roberto Treviño Cervantes. I'am building a safer, faster and more precise Search Engine, if you do not want to be part of the index, report me to rtrevinnoc@hotmail.com"
    })
    process.crawl(Indexer)
    while True:
        try:
            process.start()
        except TimeoutError:
            print("Timed out")

File: indexer.go
// Copyright (c) 2020 Roberto Treviño Cervantes

// #########################################################################
// #                                                                       #
// # This file is part of FUTURE (Powered by Monad).                       #
// #                                                                       #
// # FUTURE is free software: you can redistribute it and/or modify        #
// # it under the terms of the GNU General Public License as published by  #
// # the Free Software Foundation, either version 3 of the License, or     #
// # (at your option) any later version.                                   #
// #                                                                       #
// # FUTURE is distributed in the hope that it will be useful,             #
// # but WITHOUT ANY WARRANTY; without even the implied warranty of        #
// # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         #
// # GNU General Public License for more details.                          #
// #                                                                       #
// # You should have received a copy of the GNU General Public License     #
// # along with FUTURE.  If not, see <https://www.gnu.org/licenses/>.      #
// #                                                                       #
// #########################################################################

package main

import (
	"fmt"
	"strings"
	"strconv"
	"net/url"
	"io/ioutil"
	"math"

	"github.com/gocolly/colly/v2"
	"github.com/gocolly/colly/v2/extensions"
	"gopkg.in/neurosnap/sentences.v1/english"
	"labix.org/v2/mgo/bson"
)

func RemoveIndex(s []string, index int) []string {
	return append(s[:index], s[index+1:]...)
}

func readlines(file string) ([]string) {
	content, _ := ioutil.ReadFile(file)
	return strings.Split(string(content), "\n")
}

func loadGloveModel(file string) (map[string][]float32) {
	gloveModel := map[string][]float32{}
	literalVectors := RemoveIndex(readlines(file), 0)
	for _, literalVector := range literalVectors {
		tokens := strings.Split(literalVector, " ")
		gloveVector := []float32{}
		for index, component := range tokens {
			if index > 0 {
				preliminar, _ := strconv.ParseFloat(component, 32)
				gloveVector = append(gloveVector, float32(preliminar))
			}
		}
		gloveModel[tokens[0]] = gloveVector
	}
	return gloveModel
}

func addVectors(vectors [][]float32) ([]float32) {
	resultant := []float32{}
	for index, _ := range vectors[0] {
		var result float32
		for _, vector := range vectors {
			result += vector[index]
		}
		resultant = append(resultant, result)
	}
	return resultant
}

func divideVector(vector []float32, divisor float32) ([]float32) {
	resultant := []float32{}
	for _, component := range vector {
		resultant = append(resultant, component / divisor)
	}
	return resultant
}

func getSentenceMeanVector(sentence string) ([]float32) {
	if sentence == "" {
		return []float32{}
	}
	words := strings.Split(strings.ToLower(sentence), " ")
	wordVectors := [][]float32{}
	for _, word := range words {
		wordVectors = append(wordVectors, gloveModel[word])
	}
	return divideVector(addVectors(wordVectors), float32(len(words)))
}

func encodeURLAsNumber(url string, id int) (string) {
	text := string(id) + ":" + url
	var asciiCodesSum float64
	for index, code := range text {
		index -= len(text) + 1
		asciiCodesSum += float64(code) * math.Pow(128.0, float64(index))
	}
	return strings.Replace(strconv.FormatFloat(asciiCodesSum, 'f', -1, 64), ".", "8", 1)
}

var stopWords = readlines("stop_words.txt")
var gloveModel = loadGloveModel("glove.6B/glove.6B.50d.txt")
var tokenizer, _ = english.NewSentenceTokenizer(nil)

func saveDataFromWebPage(url url.URL, metaSiteName, metaTitle, metaDescription, webPageTitle, webPageHeader, webPageBody string) {
	hostname := url.Hostname()
	preliminarVectorElementsList := []string{}

	if metaSiteName != "" {
		preliminarVectorElementsList = append(preliminarVectorElementsList, metaSiteName)
	} else {
		preliminarVectorElementsList = append(preliminarVectorElementsList, hostname)
	}

	if metaTitle != "" {
		preliminarVectorElementsList = append(preliminarVectorElementsList, metaTitle)
	} else {
		preliminarVectorElementsList = append(preliminarVectorElementsList, webPageTitle + webPageHeader)
	}

	if metaDescription != "" {
		preliminarVectorElementsList = append(preliminarVectorElementsList, metaDescription)
	} else {
		preliminarVectorElementsList = append(preliminarVectorElementsList, webPageBody)
	}

	preliminarVectorElement := strings.Join(preliminarVectorElementsList, " ")
	sentences := tokenizer.Tokenize(preliminarVectorElement)

	for index, sentence := range sentences {
		sentence := sentence.Text

		i := strings.LastIndex(sentence, ".")
		if i > 0 {
			sentence = sentence[:i] + strings.Replace(sentence[i:], ".", "", 1)
		}
		sentence = strings.Replace(sentence, ",", "", -1)
		sentence = strings.Replace(sentence, ";", "", -1)
		sentence = strings.Replace(sentence, ":", "", -1)
		sentence = strings.Replace(sentence, "'", "", -1)
		sentence = strings.Replace(sentence, "(", "", -1)
		sentence = strings.Replace(sentence, ")", "", -1)
		sentence = strings.Replace(sentence, "[", "", -1)
		sentence = strings.Replace(sentence, "]", "", -1)
		sentence = strings.Replace(sentence, "{", "", -1)
		sentence = strings.Replace(sentence, "}", "", -1)
		sentence = strings.Replace(sentence, `"`, "", -1)
		sentence = strings.Replace(sentence, `¿`, "", -1)
		sentence = strings.Replace(sentence, `?`, "", -1)
		sentence = strings.Replace(sentence, `¡`, "", -1)
		sentence = strings.Replace(sentence, `!`, "", -1)

		webPagePointer := FUTUREIndexedWebpagePointer{
			vectorToString(getSentenceMeanVector(sentence)),
			index,
			sentence,
		}

		webPagePointerBinaryDictionary, _ := bson.Marshal(webPagePointer)
	}
}

type FUTUREIndexedWebpage struct {
	Url string
	Domain string
	Body string
	Language string

}

type FUTUREIndexedWebpagePointer struct {
	Vec string
	Vec_id int
	Sentence string
}

func vectorToString(vec []float32) (string) {
	resultant := []string{}
	for _, component := range vec {
		resultant = append(resultant, strconv.FormatFloat(float64(component), 'f', -1, 32))
	}
	return strings.Join(resultant, " ")
}

func main() {
	var metaSiteName, metaTitle, metaDescription, webPageTitle, webPageHeader, webPageBody string
	var url url.URL

	c := colly.NewCollector(
		colly.Async(true),
	)

	c.Limit(&colly.LimitRule{DomainGlob: "*", Parallelism: 10})
	extensions.RandomUserAgent(c)

	c.OnRequest(func(r *colly.Request) {
		fmt.Println("Visiting", r.URL, "as", r.Headers.Get("User-Agent"))
		url = *r.URL
	})

	c.OnXML("//meta[@property='og:description']/@content", func(e *colly.XMLElement) {
		metaDescription = strings.TrimSpace(e.Text)
	})


	c.OnXML("//meta[@property='og:title']/@content", func(e *colly.XMLElement) {
		metaTitle = strings.TrimSpace(e.Text)
	})

	c.OnXML("//meta[@property='og:description']/@content", func(e *colly.XMLElement) {
		metaSiteName = strings.TrimSpace(e.Text)
	})

	c.OnHTML("h1", func(e *colly.HTMLElement) {
		webPageHeader = strings.TrimSpace(e.Text)
	})

	c.OnHTML("title", func(e *colly.HTMLElement) {
		webPageTitle = strings.TrimSpace(e.Text)
	})

	webPageBodyElement := []string{}
	c.OnHTML("p", func(e *colly.HTMLElement) {
		webPageBodyElement = append(webPageBodyElement, e.Text)
	})
	webPageBody = strings.Join(webPageBodyElement, " ")

	c.OnHTML("img[src]", func(e *colly.HTMLElement) {
		fmt.Println("@@@@@@@@@@@@@@@@@@@@@@@@@@")
		fmt.Println(e.Attr("src"))
		fmt.Println("@@@@@@@@@@@@@@@@@@@@@@@@@@")
	})

	// Find and visit all links
	c.OnHTML("a[href]", func(e *colly.HTMLElement) {
		e.Request.Visit(e.Attr("href"))
	})

	c.OnScraped(func(r *colly.Response) {
		saveDataFromWebPage(url, metaSiteName, metaTitle, metaDescription, webPageTitle, webPageHeader, webPageBody)
	})

	c.Visit("http://rtrevinnoc.github.io/")
	c.Wait()
}
